fee-hook/src/FeeHook.sol

```diff
@@ -161,48 +237,78 @@ contract FeeHook is BaseHook, OApp  {
         bytes32 poolId) external onlyOwner
     {
         PoolId id = PoolId.wrap(poolId);
-        uint256 tokenId = TOKEN_ID[id];
+        uint256 tokenId = poolIdToTokenId[id];
         require(tokenId != 0, "Token ID not found");
         transferredPoolIds.push(poolId);
         PositionManager(payable(address(posm))).safeTransferFrom(address(this), to, tokenId, "");
     }
 
+    /**
+    * @notice Sends a request to withdraw fees, to be fulfilled by chainlink's decentralized oracle network
+    * @return requestId The ID of the request
+    */
+    function withdrawFees() external returns (bytes32 requestId) {
+        if (block.chainid != destinationChainId) {
+            revert("Withdraw: not on destination chain");
+        }
+        FunctionsRequest.Request memory req;
+        req.initializeRequestForInlineJavaScript(source); // Initialize the request with JS code
+        string memory addressString = uint256(uint160(msg.sender)).toHexString(20);
+        string[] memory args = new string[](1);
+        args[0] = addressString; // Set the address argument
+        req.setArgs(args); // Set the arguments for the request
+
+        // Send the request and store the request ID
+        requestId = _sendRequest(
+            req.encodeCBOR(),
+            subscriptionId,
+            gasLimit,
+            donID
+        );
+        requestIdToSender[requestId] = msg.sender;
 
-    function withdrawFees() external {
-        uint256 withdrawable = getWithdrawableAmount();
-        require(withdrawable > 0, "Nothing to withdraw");
-        totalWithdrawnPerUser[msg.sender] += withdrawable;
-        totalWithdrawn += withdrawable;
-        transferETH(msg.sender, withdrawable);
+        return requestId;
     }
 
-    function getWithdrawableAmount() public view returns (uint256) {
-        if (block.chainid != destinationChainId) {
-            revert("Not on destination chain. Cannot withdraw fees");
+    /**
+     * @notice Callback function for fulfilling a request to withdrawFees
+     * @param requestId The ID of the request to fulfill
+     * @param response The HTTP response data
+     * @param err Any errors from the Functions request
+     */
+    function fulfillRequest(
+        bytes32 requestId,
+        bytes memory response,
+        bytes memory err
+    ) internal override {
+        if (err.length > 0) {
+            revert(string(err));  // bubble up the actual error
         }
-        uint256 totalSupply = IERC20(vMooneyAddress).totalSupply();
-        require(totalSupply > 0, "No vMooney supply");
-        uint256 userBalance = IERC20(vMooneyAddress).balanceOf(msg.sender);
+        // Update the contract's state variables with the response and any errors
+        (uint256 totalSupply, uint256 userBalance) = abi.decode(response, (uint256, uint256));
```

Comment:
Okay... question...

Say I want to try and exploit this... could I increase my ownership right before withdrawing? Do we need to account for that? I guess they would be locking more MOONEY... but extracting current fees now, they would be increasing the total supply as well of vMOONEY right? Does the math math?

Reply:
Do we need to calculate the amount of fees each person is getting at the time that it "enters" into the contract v.s. calculate it when people are withdrawing?

e.g.: 1 ETH is generated in fees at time X: Is the relevant proportion of fees each person gets calculated at time X, or when the withdraw event happens at time X + Y when the vMOONEY distributions may have changed?

Reply:
lets put this into a separate PR, this one is already quite large.